### PR TITLE
Remove FromAsync usage from TcpClient

### PR DESCRIPTION
**Pull request by stephentoub**

### CODE REVIEW

**Comment by stephentoub** [comment]
@davidsh, could you review?  Thanks.

**Review by davidsh — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs [hunk 1]
```diff
@@ -285,23 +285,41 @@ public void Connect(IPAddress[] ipAddresses, int port)
...
-                (targetAddresses, targetPort, callback, state) => ((TcpClient)state!).BeginConnect(targetAddresses, targetPort, callback, state),
-                asyncResult => ((TcpClient)asyncResult.AsyncState!).EndConnect(asyncResult),
-                addresses, port, state: this);
+        public Task ConnectAsync(IPAddress address, int port)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this, address);
+
+            Task result = CompleteConnectAsync(Client.ConnectAsync(address, port));
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
```
Comment (scalablecory) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:294]: Are you intending `Exit` to be called prior to the connect being completed?

**Review by scalablecory — APPROVED** [review]

**Comment by davidfowl** [comment]
👏 glad to see APM usage go away.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:294]: Yeah, I'm just mirroring what the BeginConnect methods are doing below.  I'd also be perfectly happy removing all of these Enter/Exits if folks are ok with that.

**Comment by stephentoub** [comment]
> 👏 glad to see APM usage go away.

Me, too.  In spare cycles I'm aiming to remove most of the remaining usage of FromAsync, though the biggest usage is in NegotiateStream, and I'm not particularly confident in our test coverage of that right now to go rewriting everything to use async/await.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs (hunk 1 above)
Comment (davidsh) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:294]: >I'd also be perfectly happy removing all of these Enter/Exits if folks are ok with that.

Are you proposing removing the tracing completely? Or just refactoring it differently? 

I would prefer we don't remove tracing.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:294]: > Are you proposing removing the tracing completely? Or just refactoring it differently?

I'm simply not clear on the benefits of these Enter/Exit traces that we have in some (not all) methods.

Comment (davidsh) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:294]: I've been doing a lot of thinking about tracing w.r.t. to the reverse-proxy work.

These particular traces were blindly added as we ported lots of .NET Framework code to .NET Core. At the time we wanted to replicate the debug/diagnostic tracing we had with TraceListener and the debug-build GlobalLog of networking code in .NET Framework.

I agree we might want to rethink this but I would prefer we do that holistically as part of a separate PR/effort.

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:294]: > I agree we might want to rethink this but I would prefer we do that holistically as part of a separate PR/effort.

Fine by me.  Sounds like you're on top of it.  Thanks.

**Pull request merged**